proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,825 @@
+  await tryMe() // tryMe body would execute on 'preferredExecutor'
+}
+
+ await tryMe() // tryMe body would execute on 'default global concurrent executor'
+```
+
+### The `TaskExecutor` protocol
+
+In order to fulfil the requirement that we'd like default actors to run on a task executor, if it was set, we need to introduce a new kind of executor.
+
+This stems from the fact that `SerialExecutor` and how a default actor effectively acts as an executor "for itself" in Swift.
+A default actor (so an actor which does not use a custom executor), has a "default" serial executor that is created by the Swift runtime and uses the actor is the executor's identity.
+This means that the runtime executor tracking necessarily needs to track that some code is executing on a specific serial executor in order for things like `assumeIsolated` or the built-in runtime thread-safety checks can utilize them.
+
+The new protocol mirrors `Executor` and `SerialExecutor` in API, however it provides different semantics, and is tracked using a different mechanism at runtime -- by obtaining it from a task's executor preference record.
+
+The `TaskExecutor` is defined as:
+
+```swift
+public protocol TaskExecutor: Executor {
+  func enqueue(_ job: consuming ExecutorJob)
+
+  func asUnownedTaskExecutor() -> UnownedTaskExecutor
+}
+```
+
+As an intuitive way to think about `TaskExecutor` and `SerialExecutor`, one can think of the prior as being a "source of threads" to execute work on,
+and the latter being something that "provides serial isolation" and is a crucial part of Swift actors. The two share similarities, however the task executor has a more varied application space.
+
+### Task executor preference inheritance in Structured Concurrency
+
+Task executor preference is inherited by child tasks and actors which do not declare an explicit executor (so-called "default actors"), and is *not* inherited by un-structured tasks. 
+
+Specifically:
+
+* **Do** inherit task executor preference
+    * TaskGroup’s `addTask()`, unless overridden with explicit parameter
+    * `async let`
+    * methods on default actors (actors which do not use a custom executor)
+* **Do not** inherit task executor preference
+    * Unstructured tasks: `Task {}` and `Task.detached {}`
+    * methods on actors which **do** use a custom executor (including e.g. the `MainActor`)
+
+This also means that an entire tree can be made to execute their nonisolated work on a specific executor, just by means of setting the preference on the top-level task.
+
+### Task executor preference and async let
+
+Since `async let` are the simplest form of structured concurrency, they dot not offer much in the way of customization.
+
+An async currently always executes on the global concurrent executor, and with the inclusion of this proposal, it does take into account task executor preference. In other words, if an executor preference is set, it will be used by async let to enqueue its underlying task:
```

Comment:
```suggestion
An async let currently always executes on the global concurrent executor, and with the inclusion of this proposal, it does take into account task executor preference. In other words, if an executor preference is set, it will be used by async let to enqueue its underlying task:
```